File: php/lib/steam/SourceServer.php
<?php
/**
 * @author Sebastian Staudt
 * @license http://www.opensource.org/licenses/bsd-license.php Modified BSD License
 * @package Steam Interface Package (PHP)
 * @subpackage SourceServer
 * @version $Id$
 */

/**
 * @package Steam Interface Package (PHP)
 * @subpackage SourceServer
 * @todo Server has to recognize incoming packets
 */
class SourceServer
{
	/**
	 * @var int
	 */
	private $challengeNumber;
	
	/**
	 * @var mixed[]
	 */
	private $infoHash;
	
	/**
	 * @var int
	 */
	private $ping;
	
	/**
	 * @var SteamPlayer[]
	 */
	private $playerArray
	
	/**
	 * @var mixed[]
	 */
	private $rulesHash;
	
	/**
	 * @var SteamSocket
	 */
	private $socket;
	
	/**
	 * @param InetAddress $serverIP
	 * @param int $portNumber The listening port of the server, defaults to 27015
	 * @since v0.1
	 */
	public function __construct(InetAddress $ipAddress, $portNumber = 27015)
	{
		if(!is_numeric($portNumber) || $portNumber <= 0)
		{
			throw new Exception("The listening port of the server has to be a number greater than 0.");
		}
		
		$this->socket = new SteamSocket($this->ipAddress, $this->portNumber);
	}
	
	/**
   * @return The response time of this server in milliseconds
   */
	public function getPing()
	{
		return $this->ping;
	}
	
	/**
   * @return An array of SteamPlayers representing all players on this server
   */
	public function getPlayers()
	{
		return $this->playerArray;
	}
	
	/**
   * @return A associative array containing the rules of this server
   */
	public function getRules()
	{
		return $this->rulesHash;
	}
	
	/**
   * @return A associative array containing basic information about the server
   */
	public function getServerInfo()
	{
		return $this->infoHash;
	}
	
	public function initialize()
	{
		$this->updatePing();
		$this->updateServerInfo();
		$this->updateChallengeNumber();
	}
	
	/**
	 * 
	 */
	public function updatePing() 
	{
		$this->sendRequest(new A2A_PING_RequestPacket());
		$startTime = microtime(true);
		$this->getReply();
		$endTime = microtime(true);
		$this->ping = intval(round(($endTime - $startTime) * 1000));
		
		return $this->ping;
	}
	
	/**
	 * 
	 */
	public function updateChallengeNumber()
	{
		$this->sendRequest(new A2A_SERVERQUERY_GETCHALLENGE_RequestPacket());
		$this->challengeNumber = $this->getReply()->getChallengeNumber();
	}
	
 /**
   * 
   */
  public function updatePlayerInfo()
  {
    $this->sendRequest(new A2A_PLAYER_RequestPacket($this->challengeNumber));
    $this->playerArray = $this->getReply();
  }
  
  /**
   * 
   */
  public function updateRulesInfo()
  {
    $this->sendRequest(new A2A_RULES_RequestPacket($this->challengeNumber));
    $this->rulesHash = $this->getReply()->getRulesArray();
  }
	
	/**
	 * 
	 */
	public function updateServerInfo()
	{
		$this->sendRequest(new A2A_INFO_RequestPacket());
		$this->infoHash = $this->getReply()->getInfoHash();
	}
	
	/**
	 * @return SteamPacket
	 */
	private function getReply()
	{
		return $this->socket->getReply();
	}
	
	/**
	 * @param SteamPacket $requestData
	 */
	private function sendRequest(SteamPacket $requestData)
	{
		$this->socket->send($requestData);
	}
}
?>

File: ruby/lib/steam/source_server.rb
autoload "A2A_INFO_RequestPacket", "steam/packets/a2a_info_request_packet"
autoload "A2A_PING_RequestPacket", "steam/packets/a2a_ping_request_packet"
autoload "A2A_PLAYER_RequestPacket", "steam/packets/a2a_player_request_packet"
autoload "A2A_RULES_RequestPacket", "steam/packets/a2a_rules_request_packet"
autoload "A2A_SERVERQUERY_GETCHALLENGE_RequestPacket", "steam/packets/a2a_serverquery_getchallenge_request_packet"

require "steam/steam_player"
require "steam/steam_socket"

class SourceServer
  def initialize(ip_address, port_number = 27015)
    unless ip_address.is_a? IPAddr
      raise TypeError("The IP address has to be of type IPAddr")
    end
    
    unless port_number.is_a? Numeric and port_number > 0
      raise TypeError("The listening port of the server has to be a number greater than 0.")
    end
    
    @socket = SteamSocket.new ip_address, port_number
  end
  
  def get_ping
    return @ping
  end
  
  def get_players
    return @player_array
  end
  
  def get_rules
    return @rules_hash
  end
  
  def get_server_info
    return @info_hash
  end
  
  def init
    update_ping
    update_server_info
    update_challenge_number
  end

  def update_player_info
    send_request A2A_PLAYER_RequestPacket.new(@challenge_number)
    @player_array = parse_player_info get_reply
  end

  def update_rules_info
    send_request A2A_RULES_RequestPacket.new(@challenge_number)
    @rules_hash = get_reply.get_rules_hash
  end
  
  def update_server_info
    send_request A2A_INFO_RequestPacket.new
    @info_hash = get_reply.get_info_hash
  end
  
  def update_challenge_number
    send_request A2A_SERVERQUERY_GETCHALLENGE_RequestPacket.new
    @challenge_number = get_reply.get_challenge_number    
  end
 
  def update_ping
    send_request A2A_PING_RequestPacket.new
    start_time = Time.now
    get_reply
    end_time = Time.now
    return @ping = (end_time - start_time) * 1000
  end

  private

  def get_reply
    @socket.get_reply
  end
  
  def parse_player_info(player_response)
    @player_array = Array.new
    player_response.get_player_array.each do |player|
      @player_array.push SteamPlayer.new(*player)
    end
  end
  
  def send_request packet
    @socket.send packet
  end
end
